src/app/application/components/data-source/DataSource.vue

```diff
@@ -0,0 +1,81 @@
+<script setup lang="ts">
+
+import { watch, ref, onBeforeUnmount } from 'vue'
+
+import { useDataSource } from '@/utilities'
+
+const data = useDataSource()
+
+const props = defineProps({
+  src: {
+    type: String,
+    required: true,
+  },
+})
+
+const message = ref<unknown>(undefined)
+const error = ref<Error | undefined>(undefined)
+
+const emit = defineEmits<{
+  (e: 'change', value: any): void
+  (e: 'error', error: Error): void
+}>()
+
+let controller: AbortController
+let _src: string
+const sym = Symbol('')
+const open = async (src: string) => {
+  close(controller, _src)
+  _src = src
+  if (src === '') {
+    return
+  }
+  controller = new AbortController()
+  // this should emit proper events
+  const source = data.source(src, sym)
+  if (typeof source !== 'undefined') {
+    source.addEventListener(
+      'message',
+      (e) => {
+        message.value = e.data
+        emit('change', message.value)
+      },
+      { signal: controller.signal },
+    )
+    source.addEventListener(
+      'error',
+      (e) => {
+        error.value = (e as ErrorEvent).error as Error
+        emit('error', error.value)
+      },
+      { signal: controller.signal },
+    )
+  }
+}
+const close = (controller: AbortController | undefined, src: string) => {
+  if (typeof controller !== 'undefined') {
+    controller.abort()
+  }
+  if (typeof src !== 'undefined') {
+    data.close(src, sym)
+  }
+  controller = undefined
```

Comment:
either cleanup both `controller` and `_src` or neither. Doing only half is confusing

Reply:
You are shadowing global state with parameter names this makes it confusing and was the original reason for the comment above. Probably change the param name and remove the line

Reply:
I change this to use a "component global" state object with two props, rather than two separate "component global" vars. I then pass that around instead of the two vars.